// @elastic/react-search-ui/es/SearchProvider.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import PropTypes from "prop-types";
import React, { Component } from "react";
import { SearchDriver } from "@elastic/search-ui";
import SearchContext from "./SearchContext";
import defaultA11yMessages from "./A11yNotifications";
/**
 * The SearchProvider primarily holds a reference to the SearchDriver and
 * exposes it to the rest of the application in a Context.
 */

var SearchProvider = /*#__PURE__*/function (_Component) {
  _inherits(SearchProvider, _Component);

  var _super = _createSuper(SearchProvider);

  function SearchProvider() {
    var _this;

    _classCallCheck(this, SearchProvider);

    _this = _super.call(this);
    _this.state = {
      driver: null
    };
    return _this;
  }

  _createClass(SearchProvider, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var config = this.props.config; // This initialization is done inside of componentDidMount, because initializing the SearchDriver server side
      // will error out, since the driver depends on window. Placing the initialization inside of componentDidMount
      // assures that it won't attempt to initialize server side.

      var driver = new SearchDriver(_objectSpread({}, config, {
        a11yNotificationMessages: _objectSpread({}, defaultA11yMessages, {}, config.a11yNotificationMessages)
      }));
      this.setState({
        driver: driver
      });
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.state.driver.tearDown();
    }
  }, {
    key: "render",
    value: function render() {
      var children = this.props.children; // Since driver is initialized in componentDidMount above, we are waiting
      // to render until the driver is available.

      if (!this.state.driver) return null; // Passing the entire "this.state" to the Context is significant. Because
      // Context determines when to re-render based on referential identity
      // something like this could cause unnecessary renders:
      //
      // <SearchContext.Provider value={{driver: this.state.driver}}>
      //
      // By passing the entire state, we ensure that re-renders only occur when
      // state is actually updated.

      return /*#__PURE__*/React.createElement(SearchContext.Provider, {
        value: this.state
      }, children);
    }
  }]);

  return SearchProvider;
}(Component);

_defineProperty(SearchProvider, "propTypes", {
  children: PropTypes.node.isRequired,
  // Not providing a shape here because the shape matches the shape of
  // SearchDriver. SearchDriver can do it's own parameter validation.
  config: PropTypes.object
});

export default SearchProvider;
