// @elastic/react-search-ui-views/es/BooleanFacet.js
import PropTypes from "prop-types";
import React from "react";
import { FacetValue, FilterValue } from "./types";
import { appendClassName } from "./view-helpers";

function BooleanFacet(_ref) {
  var className = _ref.className,
      label = _ref.label,
      options = _ref.options,
      onChange = _ref.onChange,
      onRemove = _ref.onRemove,
      values = _ref.values;
  var trueOptions = options.find(function (option) {
    return option.value === "true";
  });
  if (!trueOptions) return null;
  var isSelected = values.includes("true");

  var apply = function apply() {
    return onChange("true");
  };

  var remove = function remove() {
    return onRemove("true");
  };

  var toggle = function toggle() {
    isSelected ? remove() : apply();
  };

  return /*#__PURE__*/React.createElement("fieldset", {
    className: appendClassName("sui-facet", className)
  }, /*#__PURE__*/React.createElement("legend", {
    className: "sui-facet__title"
  }, label), /*#__PURE__*/React.createElement("div", {
    className: "sui-boolean-facet"
  }, /*#__PURE__*/React.createElement("div", {
    className: "sui-boolean-facet__option-input-wrapper"
  }, /*#__PURE__*/React.createElement("label", {
    className: "sui-boolean-facet__option-label"
  }, /*#__PURE__*/React.createElement("div", {
    className: "sui-boolean-facet__option-input-wrapper"
  }, /*#__PURE__*/React.createElement("input", {
    className: "sui-boolean-facet__checkbox",
    type: "checkbox",
    onChange: toggle
  }), /*#__PURE__*/React.createElement("span", {
    className: "sui-boolean-facet__input-text"
  }, label)), /*#__PURE__*/React.createElement("span", {
    className: "sui-boolean-facet__option-count"
  }, trueOptions.count)))));
}

BooleanFacet.propTypes = {
  className: PropTypes.string,
  label: PropTypes.string.isRequired,
  onRemove: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(FacetValue).isRequired,
  values: PropTypes.arrayOf(FilterValue).isRequired,
  onChange: PropTypes.func.isRequired
};
export default BooleanFacet;
