// @elastic/react-search-ui/lib/SearchProvider.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _searchUi = require("@elastic/search-ui");

var _SearchContext = _interopRequireDefault(require("./SearchContext"));

var _A11yNotifications = _interopRequireDefault(require("./A11yNotifications"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

/**
 * The SearchProvider primarily holds a reference to the SearchDriver and
 * exposes it to the rest of the application in a Context.
 */
var SearchProvider = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(SearchProvider, _Component);

  var _super = _createSuper(SearchProvider);

  function SearchProvider() {
    var _this;

    (0, _classCallCheck2.default)(this, SearchProvider);
    _this = _super.call(this);
    _this.state = {
      driver: null
    };
    return _this;
  }

  (0, _createClass2.default)(SearchProvider, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var config = this.props.config; // This initialization is done inside of componentDidMount, because initializing the SearchDriver server side
      // will error out, since the driver depends on window. Placing the initialization inside of componentDidMount
      // assures that it won't attempt to initialize server side.

      var driver = new _searchUi.SearchDriver(_objectSpread({}, config, {
        a11yNotificationMessages: _objectSpread({}, _A11yNotifications.default, {}, config.a11yNotificationMessages)
      }));
      this.setState({
        driver: driver
      });
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.state.driver.tearDown();
    }
  }, {
    key: "render",
    value: function render() {
      var children = this.props.children; // Since driver is initialized in componentDidMount above, we are waiting
      // to render until the driver is available.

      if (!this.state.driver) return null; // Passing the entire "this.state" to the Context is significant. Because
      // Context determines when to re-render based on referential identity
      // something like this could cause unnecessary renders:
      //
      // <SearchContext.Provider value={{driver: this.state.driver}}>
      //
      // By passing the entire state, we ensure that re-renders only occur when
      // state is actually updated.

      return /*#__PURE__*/_react.default.createElement(_SearchContext.default.Provider, {
        value: this.state
      }, children);
    }
  }]);
  return SearchProvider;
}(_react.Component);

(0, _defineProperty2.default)(SearchProvider, "propTypes", {
  children: _propTypes.default.node.isRequired,
  // Not providing a shape here because the shape matches the shape of
  // SearchDriver. SearchDriver can do it's own parameter validation.
  config: _propTypes.default.object
});
var _default = SearchProvider;
exports.default = _default;
